doc/ApiDocGeneration/Generate-Api-Docs.ps1

```diff
@@ -0,0 +1,305 @@
+    "subscription"="Subscription";
+    "trafficmanager"="Traffic Manager"; "websites"="Websites";
+}
+
+Function Get-DocFx
+{
+    [CmdletBinding()]
+    Param ($DownloadDirectory)
+
+    try 
+    {
+        Write-Verbose "Downloading DocFx..."
+        Invoke-WebRequest -Uri "https://github.com/dotnet/docfx/releases/download/v2.43.2/docfx.zip" `
+        -OutFile "docfx.zip" | Wait-Process; Expand-Archive -Path "docfx.zip" -DestinationPath "$($DownloadDirectory)/docfx/"
+    }
+    catch 
+    {
+        Write-Verbose "Failed to Download DocFx"
+        Write-Error "$_"
+        Exit 1
+    }
+}
+
+Function Get-DocTools
+{
+    [CmdletBinding()]
+    Param (
+        $DownloadDirectory,
+        $RepoRoot
+    )
+
+    try 
+    {
+        Write-Verbose "Download and Extract mdoc to Build.BinariesDirectory/mdoc"
+        Invoke-WebRequest -Uri "https://github.com/mono/api-doc-tools/releases/download/mdoc-5.7.4.9/mdoc-5.7.4.9.zip" `
+        -OutFile "mdoc.zip" | Wait-Process; Expand-Archive -Path "mdoc.zip" -DestinationPath "$($DownloadDirectory)/mdoc/"
+
+        Write-Verbose "Restore doc/ApiDocGeneration/assets/docgen.csproj, to get ECMA2Yml and popimport"
+        dotnet restore "$($RepoRoot)/doc/ApiDocGeneration/assets/docgen.csproj" /p:BuildBinariesDirectory=$DownloadDirectory
+    }
+    catch 
+    {
+        Write-Verbose "Failed to Download Mdoc"
+        Write-Error "$_"
+        Exit 1
+    }
+
+    Get-DocFx -DownloadDirectory $DownloadDirectory
+}
+
+
+
+Function Build-Docs
+{
+    [CmdletBinding()]
+    Param (
+        $ArtifactName,
+        $ServiceDirectory,
+        $ArtifactsSafeName,
+        $ArtifactsDirectoryName,
+        $LibType,
+        $RepoRoot,
+        $BinDirectory
+    )
+
+    # Include if runing Build-Docs locally
+    # Get-DocTools -DownloadDirectory $BinDirectory -RepoRoot $RepoRoot
+
+    Write-Verbose "Create variables for identifying package location and package safe names"
+    $PackageLocation = "$($ServiceDirectory)/$($ArtifactName)"
+    Write-Verbose "Package Location $($PackageLocation)"
+    $SafeName = $ArtifactsSafeName
+
+    if ($ServiceDirectory -eq '*') {
+        $PackageLocation = "core/$($ArtifactName)"
+    }
+
+    if ($ServiceDirectory -eq 'cognitiveservices') {
+        $PackageLocation = "cognitiveservices/$($ArtifactsDirectoryName)"
+        $SafeName = $ArtifactsDirectoryName
+    }
+
+    if ($LibType -eq 'Management') {
+        $PackageLocation = "$($ServiceDirectory)/$($ArtifactName)"
+        $SafeName = $ArtifactsSafeName
+        $SafeName = $ArtifactsSafeName.Substring($ArtifactsSafeName.LastIndexOf('.Management') + 1)
+    }
+
+    Write-Verbose "Set variable for publish pipeline step"
+    echo "##vso[task.setvariable variable=artifactsafename]$($SafeName)"
+
+    Write-Verbose "Create Directories Required for Doc Generation"
+    mkdir "$($BinDirectory)/$($SafeName)/dll-docs/my-api"
```

Comment:
There is a lot of path duplication through this script we should create a set of path variables and use them through-out. It will help reduce the changes for errors if we need to tweak one of these paths.

Reply:
Excellent idea. Thanks, I have updated the script